### ISSUE TITLE
Accessibility: Installer window loses focus to a burn process

### ISSUE THREAD

**Issue by willcherry61** [wixtoolset/issues#6024]
* Which version of WiX are you building with? 
Wix Toolset v3.11.1.2318

* Which version of Visual Studio are you building with (if any)?
VS2017 version 15.9.12

* Which version of the WiX Toolset Visual Studio Extension are you building with (if any)? 0
0.9.21.62588

* Which version of .NET are you building with?
3.5

* Describe the problem and the steps to reproduce it.
The installer I am working on uses the BootstrapperApplicationRef the standard wix rtf license application. We install the vc runtime and one msi package of our own. using the MsiPackage tag in our bundle. We are having to make some adjustments to our theme.xml to fix some accessibility items.

What I have noticed is that when you install, a elevated burn process is kicked off and focus is given to that process and the ui loses focus. this prevents the Narrator app (built into windows) to read the installer ui. we've added some header text in our theme file marked as TabStop. the text says "Setup Progress" and "Setup Successful" for the different pages, one on the progress bar page, the other on the final page. These get read if I click again on the window right after hitting install, giving focus back to the UI. If I simply click install and let it run, focus is removed from the UI and the narrator stops reading. 

* Describe the behavior you expected and how it differed from the actual behavior.
I expected that the UI would keep focus, allowing the Narrator to read the pages as they change. 

see the attached video showing the issue. in the video I run Narrator, you can't hear it tho, but you can see that it highlights and puts a box around the things it does read. 

[Installer loses focus.zip](https://github.com/wixtoolset/issues/files/3259344/Installer.loses.focus.zip)

**Comment by Onkelborg** [comment]
We are currently working on replacing an NSIS installer with Wix, with a custom .NET 4.7.2 WPF bootstrapper, and I'm seeing this problem here too.

Wix 3.11.2

No difference if I start the bootstrapper from an already elevated command prompt or from a regular command prompt.

As I've understood this, the window is deactivated when the elevated engine process is being started.

I'm not sure if I'm doing something wrong, and if so what, or if it's a bug. I've peeked at the sources a bit for burn, and my best guess is the call to ShellExecuteEx?

https://github.com/wixtoolset/wix3/blob/2ed6d7906c87bb65b75cfa0020e38a1b1dffa49b/src/burn/engine/pipe.cpp#L423
`
hr = ShelExec(wzExecutablePath, sczParameters, wzVerb, NULL, SW_HIDE, hwndParent, &hProcess);
`

I'm thinking about the SW_HIDE parameter: https://docs.microsoft.com/en-us/windows/win32/api/shellapi/nf-shellapi-shellexecutea#sw_hide-0
> Hides the window and activates another window.

I'm not sure how to interpret this parameter..?

**Comment by Onkelborg** [comment]
Ok, I think I found a (partial) solution: Calling LockSetForegroundWindow before the elevation to lock, and after to unlock, seems to resolve the issue. I only made these calls from my managed bootstrapper, however, I guess they would make more sense to do inside the engine instead?

**Comment by willcherry61** [comment]
Ya, makes sense to me. If the engine makes it, I don't have to worry about it.

Get Outlook for Android<https://aka.ms/ghei36>
________________________________

**Comment by rseanhall** [comment]
The "SW" in `SW_HIDE` stands for [ShowWindow](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-showwindow), which says:

> Controls how the window is to be shown. This parameter is ignored the first time an application calls ShowWindow, if the program that launched the application provides a STARTUPINFO structure. Otherwise, the first time ShowWindow is called, the value should be the value obtained by the WinMain function in its nCmdShow parameter. In subsequent calls, this parameter can be one of the following values.

So it controls the `nCmdShow` parameter of the first time that the elevated process calls `ShowWindow`, which should never happen because the elevated process doesn't show UI.

These kinds of bugs are normally resolved by finding the relevant blog post from https://devblogs.microsoft.com/oldnewthing. I'm not seeing one though, which is probably why this hasn't been fixed.

It's not clear why the elevated process is causing the window to be deactivated. Ideally we could prevent that happening in the first place, because `LockSetForegroundWindow` probably will also fail in some circumstances.

**Comment by willcherry61** [comment]
Gotcha, so is that something I would set in our installer?

**Comment by rseanhall** [comment]
If you're using a built-in BA, then no you shouldn't have to do anything since we would make the standard BA do what's necessary if the engine can't do it.

**Comment by rseanhall** [comment]
If someone is interested in fixing this, the root cause is probably that the hidden window that the engine creates for broadcast messages is not actually a hidden window:
v3 - https://github.com/wixtoolset/wix3/blob/2ed6d7906c87bb65b75cfa0020e38a1b1dffa49b/src/burn/engine/uithread.cpp#L137
v4 - https://github.com/wixtoolset/burn/blob/422d6e48e2f9bdcec62f9a147cb8c2f34bf73a5f/src/engine/uithread.cpp#L137

**Comment by BMurri** [comment]
I'd like to fix this

**Comment by Onkelborg** [comment]
That window shouldn't need to be visible,, it's just to be able to receive window messages, right..?

**Comment by BMurri** [comment]
@Onkelborg it does need to receive broadcast messages from the system, so it must be a top-level window and cannot be a message window. Other than that, I'm just verifying that it isn't made easily available externally so that no one tries to make it a parent window, since it's intended to never be shown.

It was always intended to be hidden, but I assume someone copied their broilerplate into the code and it made it past reviews because it looked so normal.

**Comment by barnson** [comment]
It has to be visible to get shutdown messages.

**Comment by rseanhall** [comment]
If it needs to be visible, then it needs to properly show it without activating it. The [CreateWindowEx documentation](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-createwindowexw) says that the `CW_USEDEFAULT` value for `x` (with the `SW_*` value in `y`) is not valid unless it's an overlapped window. `WS_EX_TOOLWINDOW` is not an overlapped window. Adding the extended window style `WS_EX_NOACTIVATE` is probably also a good idea.

**Comment by BMurri** [comment]
According to the [Window Features](https://docs.microsoft.com/en-us/windows/win32/winmsg/window-features) documentation:

- a _message-only_ window does not receive broadcast messages, so it's no use to us in this instance.
- only windows _parented_ by the desktop (aka top-level) receive broadcast messages, so we must not be a child window.
- disabled windows _cannot_ receive user input, but it _can_ receive messages from other windows, other applications, and the system (aka broadcast messages), so we can be disabled.
- hidden windows are _effectively_ disabled. Nothing in the **Window Visibility** section says anything about messages not related to user input not being received by hidden windows.

So, the window doesn't need to be visible according to this page. But. to be certain, I tried it out anyway.

With the same changes in v3 and v4, I got the following line from the log as I rebooted the box while burn was waiting for input (line is from v3, but v4 is similar):

`[2238:19D8][2021-03-04T21:11:39]i400: Received system request to shut down the process: critical: No, elevated: No, allowed: Yes`

Seems to work just fine. I'm pushing pull requests out soon,

**Comment by rseanhall** [comment]
It looks like it's a little more nuanced than just being visible: https://docs.microsoft.com/en-us/previous-versions/windows/desktop/ms700677(v=vs.85)#certain-types-of-applications-will-no-longer-be-allowed-to-block-shutdown. In Vista+, we need to be visible in order to be able to block shutdown or we need to call [ShutdownBlockReasonCreate](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-shutdownblockreasoncreate).

**Comment by BMurri** [comment]
Even if we are visible, the "reason" string is something we probably should do anyway. But, if we use that API, we are treated as if we were visible even if we aren't.

I'm looking at what would be the best way to incorporate this new API. I'll follow up on wixdevs.

**Comment by rseanhall** [comment]
I would prefer not to use `ShutdownBlockReasonCreate` to fix this bug, especially if you really want this in v3. That seems like a separate feature request to me.

**Comment by BMurri** [comment]
As I'm looking at what is involved, WRT v3, I agree.

I think it's a good idea for v4, though I may include it as part of issue #6349 (haven't determined that, yet).
